primitives/xcm/src/lib.rs

```diff
@@ -344,3 +353,75 @@ impl<
 		}
 	}
 }
+
+// Struct that converts a given MultiLocation into a 20 bytes account id by hashing
+// with blake2_256 and taking the first 20 bytes
+pub struct Account20Hash<AccountId>(PhantomData<AccountId>);
+impl<AccountId: From<[u8; 20]> + Into<[u8; 20]> + Clone> Convert<MultiLocation, AccountId>
+	for Account20Hash<AccountId>
+{
+	fn convert_ref(location: impl Borrow<MultiLocation>) -> Result<AccountId, ()> {
```

Comment:
Is this `MultiLocation` sufficiently normalized? If there were two different ways to represent the same ML, this could result in different `AccountId`s which represent the same location.

Reply:
Also, how future-proof is this? If we get a new ML version, would this hash to a different account (and how would we handle that)?

Reply:
This is a very good question. First, yes MultiLocation should be sufficiently normalized (in the sense that in the Moonbeams view each account in each parachain will correspond to a different multilocation).

As for the future multilocation versions. First, xcm v3 does not change it. But assuming a future version does changes this, we have 2 distinct cases:

- Not a breaking change: if its not a breaking change, we can always convert back to V2 MultiLocations, and work with these while we give some time to migrate the funds to the new hash. This should be easily doable with a XCM message.

- Breaking change: Same thing applies I guess, but **we would have to offer this solutions before we upgrade to the new version**

Reply:
So according to Parity we shouldnt have issues with this, as we should be able to convert from subsequent versions to the current one. I think we should proceed forward with this given their response